lib/id3c/cli/command/geocode.py

```diff
@@ -0,0 +1,435 @@
+        # config file.
+        if config_dir:
+            chdir(config_dir)
+
+    for config in configs:
+        try:
+            kwargs = {
+                "filename":          config["filename"],
+                "street_column":     config["columns"]["street"],
+                "city_column":       config["columns"]["city"],
+                "state_column":      config["columns"]["state"],
+                "zipcode_column":    config["columns"]["zipcode"],
+                "secondary_column":  config["columns"].get("secondary")
+            }
+        except KeyError as key:
+            LOG.error(f"Required key «{key}» missing from config {config}")
+            raise key from None
+
+        geocoded_addresses = get_geocoded_addresses(**kwargs)
+        geocoded_addresses.to_csv(sys.stdout, index=False)
+
+
+def get_geocoded_addresses(*,
+                           filename,
+                           street_column,
+                           city_column,
+                           state_column,
+                           zipcode_column,
+                           secondary_column = None):
+    """
+    Internal function powering :func: `geocode_using_options` and
+    :func: `geocode_using_config`.
+    """
+    LOG.debug(f"Reading addresses file {filename}")
+
+    address_file_df = load_file_as_dataframe(filename)
+
+    address_column_map = {
+        'street': street_column,
+        'city': city_column,
+        'state': state_column,
+        'zipcode': zipcode_column,
+        'secondary': secondary_column
+    }
+
+    addresses_df = create_standardized_address_df(address_file_df, address_column_map)
+    standardized_addresses = convert_df_to_dict_series(addresses_df)
+
+    cache = load_or_create_cache()
+
+    smartystreets_responses = pd.DataFrame(columns = ['lat','lng','canonicalized_address'])
+
+    for address in standardized_addresses:
+        response = get_response_from_cache_or_geocoding(address, cache)
+
+        smartystreets_responses = smartystreets_responses \
+            .append(response,ignore_index = True)
+
+    geocoded_addresses = address_file_df.join(smartystreets_responses)
+    save_cache(cache)
+
+    return geocoded_addresses
+
+
+def load_file_as_dataframe(filename: str) -> pd.DataFrame:
+    """
+    Given a *filename*, loads its data as a pandas DataFrame.
+
+    Raises a :class: `UnsupportedFileExtensionError` if the given *filename*
+    ends with an unsupported extension.
+    """
+    if not filename.endswith(SUPPORTED_EXTENSIONS):
+        raise UnsupportedFileExtensionError(dedent(f"""
+            Unsupported file extension for file «{filename}».
+            Please choose from one of the following file extensions:
+                {SUPPORTED_EXTENSIONS}
+            """
+        ))
+
+    if filename.endswith(('.csv', '.tsv')):
+        separator = '\t' if filename.endswith('.tsv') else ','
+        df = pd.read_csv(filename, sep=separator, dtype = str)
```

Comment:
this is a nitpick but I prefer the style of using no spaces around named function arguments. Either is fine but there's a mix of each in this line of code.